src/solidity/decoders.js

```diff
@@ -0,0 +1,387 @@
+    var dim = typeName.match(squareBracket)
+    var size = dim[dim.length - 1]
+    if (size === '[]') {
+      return 'dynamic'
+    } else {
+      return parseInt(dim[dim.length - 1].replace('[', '').replace(']'))
+    }
+  }
+}
+
+/**
+  * get the decoder which decode the sub array of the current @arg arrayType
+  *
+  * @param {String} typeName - short type ( e.g uint[][4] )
+  * @param {Object} variable - variable given by the AST
+  * @param {Object} stateDefinitions - all state stateDefinitions given by the AST (including struct and enum type declaration)
+  * @return {Object} return the array decoder
+  */
+function getSubArrayDecoder (typeName, variable, stateDefinitions) {
+  var subArray = null
+  var subArrayType = typeName.substring(0, typeName.lastIndexOf('['))
+  if (subArrayType.indexOf('[') !== -1) {
+    subArray = JSON.parse(JSON.stringify(variable))
+    subArray.attributes.type = subArrayType
+    subArray = decode(subArray, stateDefinitions)
+  }
+  return subArray
+}
+
+/**
+  * extract the underlying type
+  *
+  * @param {String} fullType - type given by the AST (ex: uint[2] storage ref[2])
+  * @return {String} return the first part of the full type. don not keep the array declaration ( uint[2] storage ref[2] will return uint)
+  */
+function extractUnderlyingType (fullType) {
+  var splitted = fullType.split(' ')
+  if (fullType.indexOf('enum') === 0 || fullType.indexOf('struct') === 0) {
+    return splitted[0] + ' ' + splitted[1]
+  }
+  if (splitted.length > 0) {
+    fullType = splitted[0]
+  }
+  if (fullType[fullType.length - 1] === ']') {
+    return fullType.substring(0, fullType.indexOf('['))
+  }
+  return fullType
+}
+
+var decoders = {
+  'address': Address,
+  'array': ArrayType,
+  'bool': Bool,
+  'bytes': DynamicByteArray,
+  'bytesX': FixedByteArray,
+  'enum': Enum,
+  'string': StringType,
+  'struct': Struct,
+  'int': Int,
+  'intX': Int,
+  'uint': Uint,
+  'uintX': Uint
+}
+
+/**
+  * parse the full type
+  *
+  * @param {String} fullType - type given by the AST (ex: uint[2] storage ref[2])
+  * @return {String} returns the token type (used to instanciate the right decoder) (uint[2] storage ref[2] will return 'array', uint256 will return uintX)
+  */
+function extractTokenType (fullType) {
+  if (fullType.indexOf('[') !== -1) {
+    return 'array'
+  }
+  if (fullType.indexOf(' ') !== -1) {
+    fullType = fullType.split(' ')[0]
+  }
+  return fullType.replace(/[0-9]+/g, 'X')
+}
+
+/**
+  * parse the type and return an object representing the type
+  *
+  * @param {Object} variable - variable declaration given by the AST
+  * @param {Object} stateDefinitions - all state stateDefinitions given by the AST (including struct and enum type declaration)
+  * @return {Object} - return the corresponding decoder
+  */
+function decode (variable, stateDefinitions) {
+  var currentType = extractTokenType(variable.attributes.type)
+  return new decoders[currentType](variable, stateDefinitions)
```

Comment:
This only needs `variable.attributes.type`, actually this whole module only needs the type string.